pyro/contrib/autoguide/__init__.py

```diff
@@ -297,6 +294,59 @@ def median(self, *args, **kwargs):
         return self(*args, **kwargs)
 
 
+def _hessian(y, xs):
+    dys = torch.autograd.grad(y, xs, create_graph=True)
+    flat_dy = torch.cat([dy.reshape(-1) for dy in dys])
+    H = []
+    for dyi in flat_dy:
+        Hi = torch.cat([Hij.reshape(-1) for Hij in torch.autograd.grad(dyi, xs, retain_graph=True)])
```

Comment:
Yup, what I understand from autograd's docs is that: With `create_graph` enable, backward will build the graph to compute gradients from inputs; for second-order derivatives, we don't need to enable `create_graph` because we don't want to compute third-order derivatives here. However, we have to enable `retain_graph` because we will use the graph again and again for each `dyi`.

Reply:
It seems that things will be freed if we don't use `retain_graph`.
```
x = torch.rand(3, requires_grad=True)
y = x ** 2
z = torch.autograd.grad(y[0], x)
t = torch.autograd.grad(y[1], x)  # <- RuntimeError
```
We get RuntimeError said that "Trying to backward from the graph second time, but the buffers have already been freed. Specify `retain_graph=True` when calling backward the first time".

Reply:
Thanks for explaining!